python/ql/src/experimental/dataflow/internal/DataFlowPrivate.qll

```diff
@@ -389,20 +395,11 @@ string ppReprType(DataFlowType t) { none() }
  * taken into account.
  */
 predicate jumpStep(Node nodeFrom, Node nodeTo) {
-  // As we have ESSA variables for global variables,
-  // we include ESSA flow steps involving global variables.
-  (
-    nodeFrom.(EssaNode).getVar() instanceof GlobalSsaVariable
-    or
-    nodeTo.(EssaNode).getVar() instanceof GlobalSsaVariable
-  ) and
+  EssaFlow::essaFlowStep(nodeFrom, nodeTo) and
   (
-    EssaFlow::essaFlowStep(nodeFrom, nodeTo)
+    nodeFrom instanceof ModuleVariableNode
     or
-    // As jump steps do not respect chronology,
-    // we add jump steps for each def-use pair.
-    nodeFrom.asVar() instanceof GlobalSsaVariable and
-    nodeTo.asCfgNode() = nodeFrom.asVar().getASourceUse()
+    nodeTo instanceof ModuleVariableNode
   )
 }
```

Comment:
The fact that we only allow jumpStep with a global variable that escapes (`ModuleVariableNode`) seems to be a problem, and what is causing us to loose flow.


`python/ql/test/experimental/dataflow/basic/test.py` contains:

```py
def obfuscated_id(x):
  y = x
  z = y
  return z

a = 42
b = obfuscated_id(a)
```

And [used to have flow](https://github.com/github/codeql/pull/4265/files#diff-87b373ad8be2edb3f84a91371c4de57eL31-L52) from `a -> x`, `a -> y`, `a -> z`, `a -> b`. But since `a` doesn't escape, we won't include it anymore.

Reply:
I agree with your assessment, but I don't think allowing non-escaping variables is the solution. Because jump steps ignore chronology, something like the following would have flow from source to sink if `foo` has a `ModuleVariableNode`:
```python
foo = NON_SOURCE
SINK(foo)
foo = SOURCE
```
Now, I was hoping that the necessary flow in the above would fall out from just the regular SSA flow, but it seems it doesn't.

Note that if `foo` escapes, then we _do_ get the bad flow above anyway, and this is in some sense a bit spurious again (since the second line will never execute a second time with `foo` containing `SOURCE`). One solution would be to not consider all top-level reads and writes to global variables, but only the ones that ultimately define the variable in question.

Here's another problematic test example:
```python
foo = SOURCE
foo = NON_SOURCE
SINK(foo)
```
Currently we would have flow from source to sink if `foo` is considered to be escaping.

Either way, I think we need to have "local" Global SSA flow for modelling assignments during module initialisation, and flow to and from module variable nodes to model steps via reads and writes of globals _outside_ of initialisation.